content/blog/2022-05-19-A_Type_and_Effect_System_for_Deterministic_Parallel_Java/index.md

```diff
@@ -0,0 +1,123 @@
+
+The code below shows an example of a class definition (a Tree class for a tree data structure) with annotations. The example is taken from the original paper, and we added more comments in it to show the functionality.
+
+```
+// take the "global" region P for this class as a template argument.
+class TreeNode<region P> {
+     region Links, L, R;  // create DPJ regions
+     // annotate that the member variable `mass` is in region P.
+     double mass in P ;
+     // annotate that member variables `left` and `right` are in region `Links`.
+     TreeNode<L> left in Links;
+     TreeNode<R> right in Links ;
+     // Annotate effect for the function `setMass` saying that it writes to region P.
+     void setMass(double mass) writes P { this.mass = mass; }
+     void initTree(double mass) {
+       cobegin {
+         // overal effect: reads Links writes L
+         left.mass = mass;
+         // overal effect: reads Links writes R
+         right.mass = mass;
+       }
+    }
+ }
+```
+
+### Region Nesting
+
+Since many datastructures are naturally recursive, DPJ comes with the functionality to express recursiveness of memory regions. This is called **region nesting**. The listing below shows how the aforementioned program can be extended with region nesting:
+
+```
+// take the "global" region P for this class as a template argument.
+class TreeNode<region P> {
+     region Links, L, R;  // create DPJ regions
+     // annotate that the member variable `mass` is in region P.
+     double mass in P:M ; // add P:M showing nesting of M under P
+     // annotate that member variables `left` and `right` are in region `Links`.
+     TreeNode<P:L> left in Links; // add P:L showing nesting of L under P
+     TreeNode<P:R> right in Links ; // add P:R showing nesting of R under P
+     // Annotate effect for the function `setMass` saying that it writes to region P.
+     void setMass(double mass) writes P { this.mass = mass; }
+     void initTree(double mass) {
+       cobegin {
+         // overal effect: reads Links writes L
+         left.mass = mass;
+         // overal effect: reads Links writes R
+         right.mass = mass;
+       }
+    }
+ }
+```
+
+The above code results in the following recursive region nesting as shown in the figure below. Here, the mass region M is different for every nested/recursive node.
+
+<p align="center">
+<img src="figure_1.jpg" alt="alt_text" title="image_tooltip" width="500" />
+</p>
+
+### Subtyping
+
+The DPJ type system supports subtyping to better express nested regions. A summary of the subtyping rules is shown below:
+
+* Two regions are nested within each other, R1 ≤ R2,  iff R1 is contained within R2, e.g. P:M ≤ P;
+* C<R1> is a subtype of C<R2> iff every region of R1 is in R2 (R1 ⊆ R2); examples:
+C<P:M> is a subtype of C<P:*>
+C<P:L:*:M> is a subtype of C<P:*:M>
```

Comment:
Some of these asterisks have turned into italics formatting in Markdown. Using math mode can help avoid this, of course. But you will want to preview your post using Zola to make sure it looks right.